robotiq_action_server/src/robotiq_c_model_action_server.cpp

```diff
@@ -0,0 +1,194 @@
+    result.rSP = 128; // Middle ground speed
+    
+    if (goal.command.position > params.max_gap_ || goal.command.position < params.min_gap_)
+    {
+      ROS_WARN("Goal gripper gap size is out of range(%f to %f): %f m",
+               params.min_gap_, params.max_gap_, goal.command.position);
+      throw BadArgumentsError();
+    }
+    
+    if (goal.command.max_effort < params.min_effort_ || goal.command.max_effort > params.max_effort_)
+    {
+      ROS_WARN("Goal gripper effort out of range (%f to %f N): %f N",
+               params.min_effort_, params.max_effort_, goal.command.max_effort);
+      throw BadArgumentsError();
+    }
+
+    double dist_per_tick = (params.max_gap_ - params.min_gap_) / 255;
+    double eff_per_tick = (params.max_effort_ - params.min_effort_) / 255;
+
+    result.rPR = static_cast<uint8_t>((params.max_gap_ - goal.command.position) / dist_per_tick);
+    result.rFR = static_cast<uint8_t>((goal.command.max_effort - params.min_effort_) / eff_per_tick);
+
+    ROS_INFO("Setting goal position register to %hhu", result.rPR);
+
+    return result;
+  }
+
+  /*  This function is templatized because both GripperCommandResult and GripperCommandFeedback consist
+      of the same fields yet have different types. Templates here act as a "duck typing" mechanism to avoid
+      code duplication.
+  */
+  template<typename T>
+  T registerStateToResultT(const GripperInput& input, const CModelGripperParams& params, uint8_t goal_pos)
+  {
+    T result;
+    double dist_per_tick = (params.max_gap_ - params.min_gap_) / 255;
+    double eff_per_tick = (params.max_effort_ - params.min_effort_) / 255;
+
+    result.position = input.gPO * dist_per_tick + params.min_gap_;
+    result.effort = input.gCU * eff_per_tick + params.min_effort_;
+    result.stalled = input.gOBJ == 0x1 || input.gOBJ == 0x2;
+    result.reached_goal = input.gPO == goal_pos;
+
+    return result;
+  }
+
+  // Inline api-transformers to avoid confusion when reading the action_server source
+  inline
+  GripperCommandResult registerStateToResult(const GripperInput& input, const CModelGripperParams& params, uint8_t goal_pos)
+  {
+    return registerStateToResultT<GripperCommandResult>(input, params, goal_pos);
+  }
+
+  inline
+  GripperCommandFeedback registerStateToFeedback(const GripperInput& input, const CModelGripperParams& params, uint8_t goal_pos)
+  {
+    return registerStateToResultT<GripperCommandFeedback>(input, params, goal_pos);
+  }
+
+} // end of anon namespace
+
+
+ras::CModelGripperActionServer::CModelGripperActionServer(const std::string& name, const CModelGripperParams& params)
+  : nh_()
+  , as_(nh_, name, false)
+  , action_name_(name)
+  , gripper_params_(params)
+{
+  as_.registerGoalCallback(boost::bind(&CModelGripperActionServer::goalCB, this));
+  as_.registerPreemptCallback(boost::bind(&CModelGripperActionServer::preemptCB, this));
+
+  state_sub_ = nh_.subscribe(name + "/input", 1, &CModelGripperActionServer::analysisCB, this);
```

Comment:
Creating multiple instances within a single node is not typical.  It also means that remapping cannot be done easily.  It would require intimate knowledge about how names are concatenated.  Instead of creating multiple instances in a single process, launching multiple nodes is preferred.